cloud: support chaining of assumed roles

rhu713 opened the pull request:
Previously, a user could only assume a role in AWS or GCP directly, via the
identity specified by implicit or specified auth. This was inadequate because
there is a need to support role assumption through a chain of intermediate
delegate roles. To address this, this patch allows the ASSUME_ROLE parameter
in AWS and GCP storage and KMS URIs to specify a list of roles with a
comma-separated string. The roles in the list can then be chain assumed in
order to access the resource specified by the URI.

With this patch, if a destination in S3 can only be accessed by RoleB, and the
identity corresponding to implicit auth can only assume RoleB through an
intermediate role RoleA, then this chain assumption can be specified in the S3
URI:
s3://bucket/key?AUTH=implicit&ASSUME_ROLE=RoleA,RoleB

In addition, the parameters for auth in AWS URIs via assume role have been
changed so that the "assume" auth mode no longer exists, and the ASSUME_ROLE
param can be specified for both "specified" and "implicit" auth.

Finally, some AWS cloud unit tests, including the assume role tests, have been
added to the unit test nightly.

Release note (enterprise change): Allow the ASSUME_ROLE parameter in AWS and
GCP storage and KMS URIs to specify a list of roles with a comma-separated
string. The roles in the list can then be chain assumed in order to access the
resource specified by the URI.

For example, if a destination in S3 can only be accessed by RoleB, and the
identity corresponding to implicit auth can only assume RoleB through an
intermediate role RoleA, then this chain assumption can be specified in the S3
URI:
s3://bucket/key?AUTH=implicit&ASSUME_ROLE=RoleA,RoleB

In addition, remove the "assume" auth mode from AWS URIs, and instead use the
ASSUME_ROLE parameter to indicate that a role should be assumed for
authentication. Below are some examples:

S3: s3://<bucket>/<key>?AUTH=specified&ASSUME_ROLE=<role_arn>&AWS_ACCESS_KEY_ID=<access_key>&AWS_SECRET_ACCESS_KEY=<secret_key>
AWS KMS: aws:///<key_arn>?AUTH=implicit&REGION=<region>&ASSUME_ROLE=<role_arn>

Addresses https://github.com/cockroachdb/cockroach/issues/81453

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83712)
<!-- Reviewable:end -->

Commit:
cloud: support chaining of assumed roles

Previously, a user could only assume a role in AWS or GCP directly, via the
identity specified by implicit or specified auth. This was inadequate because
there is a need to support role assumption through a chain of intermediate
delegate roles. To address this, this patch allows the ASSUME_ROLE parameter
in AWS and GCP storage and KMS URIs to specify a list of roles with a
comma-separated string. The roles in the list can then be chain assumed in
order to access the resource specified by the URI.

With this patch, if a destination in S3 can only be accessed by RoleB, and the
identity corresponding to implicit auth can only assume RoleB through an
intermediate role RoleA, then this chain assumption can be specified in the S3
URI:
s3://bucket/key?AUTH=implicit&ASSUME_ROLE=RoleA,RoleB

In addition, the parameters for auth in AWS URIs via assume role have been
changed so that the "assume" auth mode no longer exists, and the ASSUME_ROLE
param can be specified for both "specified" and "implicit" auth.

Finally, some AWS cloud unit tests, including the assume role tests, have been
added to the unit test nightly.

Release note (enterprise change): Allow the ASSUME_ROLE parameter in AWS and
GCP storage and KMS URIs to specify a list of roles with a comma-separated
string. The roles in the list can then be chain assumed in order to access the
resource specified by the URI.

For example, if a destination in S3 can only be accessed by RoleB, and the
identity corresponding to implicit auth can only assume RoleB through an
intermediate role RoleA, then this chain assumption can be specified in the S3
URI:
s3://bucket/key?AUTH=implicit&ASSUME_ROLE=RoleA,RoleB

In addition, remove the "assume" auth mode from AWS URIs, and instead use the
ASSUME_ROLE parameter to indicate that a role should be assumed for
authentication. Below are some examples:

S3: s3://<bucket>/<key>?AUTH=specified&ASSUME_ROLE=<role_arn>&AWS_ACCESS_KEY_ID=<access_key>&AWS_SECRET_ACCESS_KEY=<secret_key>
AWS KMS: aws:///<key_arn>?AUTH=implicit&REGION=<region>&ASSUME_ROLE=<role_arn>

adityamaru commented on pkg/cloud/gcp/gcs_storage.go:
```diff
@@ -44,8 +44,10 @@ const (
 	// the Google Application Credentials JSON file.
 	CredentialsParam = "CREDENTIALS"
-	// AssumeRoleParam is the email of the service account to assume.
+	// AssumeRoleParam is query parameter for the chain service account email
```
nit: is the query parameter

rhu713 replied:
done

adityamaru commented on pkg/cloud/amazon/aws_kms_test.go:
```diff
@@ -57,73 +58,67 @@ func TestEncryptDecryptAWS(t *testing.T) {
-	kmsRegion := os.Getenv("AWS_KMS_REGION_A")
+	kmsRegion := os.Getenv("AWS_KMS_REGION")
 	if kmsRegion == "" {
 		skip.IgnoreLint(t, "AWS_KMS_REGION_A env var must be set")
```
nit: change to AWS_KMS_REGION

rhu713 replied:
done

adityamaru commented on pkg/cloud/amazon/s3_storage.go:
```diff
@@ -70,9 +71,8 @@ const (
-	// AWSRoleArnParam is the query parameter for the 'role ARN' in an AWS
-	// URI, use if credentials are being verified with Assume Role.
-	AWSRoleArnParam = "AWS_ROLE_ARN"
+	// AssumeRoleParam is query parameter for the chain AWS Role ARNs to assume.
```
nit: is the query parameter

adityamaru replied:
nit: for the chain of AWS Role ARNs?

rhu713 replied:
done

adityamaru commented on pkg/cloud/amazon/s3_storage.go:
```diff
@@ -70,9 +71,8 @@ const (
-	// URI, use if credentials are being verified with Assume Role.
-	AWSRoleArnParam = "AWS_ROLE_ARN"
+	// AssumeRoleParam is query parameter for the chain AWS Role ARNs to assume.
+	AssumeRoleParam = "ASSUME_ROLE"
```
just sanity checking, we havent backported any of this to 22.1 right, in which case its okay to switch the param name.

adityamaru commented on pkg/cloud/amazon/s3_storage_test.go:
```diff
@@ -206,19 +206,56 @@ func TestPutS3AssumeRole(t *testing.T) {
+		// Finally, check that the chain of roles can be used to access the storage.
+		uri := S3URI(bucket, "backup-test",
+			&roachpb.ExternalStorage_S3{
+				Auth:             cloud.AuthParamSpecified,
```
ditto as below, worth adding an implicit version of this test?

adityamaru reviewed (changes requested):
Mostly LGTM! Just a few nits and a question about testing IMPLICT + chaining everywhere. Thanks for adding the AWS tests to the nightly as well 🙌

rhu713 commented on pkg/cloud/amazon/aws_kms.go:
```diff
@@ -169,8 +147,23 @@ func MakeAWSKMS(ctx context.Context, uri string, env cloud.KMSEnv) (cloud.KMS, e
-	if kmsURIParams.auth == cloud.AuthParamAssume {
-		sess.Config.Credentials = stscreds.NewCredentials(sess, kmsURIParams.roleArn)
+	if kmsURIParams.roleARN != "" {
+		for _, role := range kmsURIParams.delegateRoleARNs {
```
done

rhu713 commented on pkg/cloud/gcp/gcs_storage_test.go:
```diff
@@ -213,6 +214,49 @@ func TestGCSAssumeRole(t *testing.T) {
+		roleChain := strings.Split(roleChainStr, ",")
+
+		q := make(url.Values)
+		q.Set(cloud.AuthParam, cloud.AuthParamSpecified)
```
done, now testing both specified and implicit auth.

rhu713 commented:
bors r+

rhu713 commented:
bors retry
